src/azure-cli/azure/cli/command_modules/util/custom.py

```diff
@@ -175,6 +175,51 @@ def upgrade_version(cmd, update_all=None, yes=None):  # pylint: disable=too-many
                    else auto_upgrade_msg)
 
 
+def _upgrade_on_windows():
+    """Download MSI to a temp folder and install it with msiexec.exe.
+    Directly installing from URL may be blocked by policy: https://github.com/Azure/azure-cli/issues/19171
+    This also gives the user a chance to manually install the MSI in case of msiexec.exe failure.
+    """
+    logger.warning("Updating Azure CLI with MSI from https://aka.ms/installazurecliwindows")
+    tmp_dir, msi_path = _download_from_url('https://aka.ms/installazurecliwindows')
+
+    logger.warning("Installing MSI")
+    import subprocess
+    exit_code = subprocess.call(['msiexec.exe', '/i', msi_path])
+
+    if exit_code:
+        logger.warning("Installation Failed. You may manually install %s", msi_path)
+    else:
+        from azure.cli.core.util import rmtree_with_retry
+        logger.warning("Succeeded. Deleting %s", tmp_dir)
+        rmtree_with_retry(tmp_dir)
+    return exit_code
+
+
+def _download_from_url(url):
+    import requests
+    from azure.cli.core.util import should_disable_connection_verify
+    r = requests.get(url, stream=True, verify=(not should_disable_connection_verify()))
+    if r.status_code != 200:
+        raise CLIError("Request to {} failed with {}".format(url, r.status_code))
+
+    # r.url is the real path of the msi, like'https://azcliprod.blob.core.windows.net/msi/azure-cli-2.27.1.msi'
+    file_name = r.url.rsplit('/')[-1]
+    import tempfile
+    tmp_dir = tempfile.mkdtemp()
+    msi_path = os.path.join(tmp_dir, file_name)
+    logger.warning("Downloading MSI to %s", msi_path)
```

Comment:
Similarly, isn't this more of a `logger.debug`?

Reply:
Same as above.